cookie_source.go

```diff
@@ -0,0 +1,73 @@
+package mint
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/hkdf"
+)
+
+// CookieSource is used to create and verify source address tokens
+type CookieSource interface {
+	// NewToken creates a new token
+	NewToken([]byte) ([]byte, error)
+	// DecodeToken decodes a token
+	DecodeToken([]byte) ([]byte, error)
+}
+
+type defaultCookieSource struct {
+	aead cipher.AEAD
+}
+
+const tokenKeySize = 16
+const tokenNonceSize = 16
+
+// newDefaultCookieSource creates a source for source address tokens
+func newDefaultCookieSource() (CookieSource, error) {
+	secret := make([]byte, 32)
+	if _, err := rand.Read(secret); err != nil {
+		return nil, err
+	}
+	key, err := deriveKey(secret)
+	if err != nil {
+		return nil, err
+	}
+	c, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	aead, err := cipher.NewGCMWithNonceSize(c, tokenNonceSize)
+	if err != nil {
+		return nil, err
+	}
+	return &defaultCookieSource{aead: aead}, nil
+}
+
+func (s *defaultCookieSource) NewToken(data []byte) ([]byte, error) {
+	nonce := make([]byte, tokenNonceSize)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	return s.aead.Seal(nonce, nonce, data, nil), nil
```

Comment:
I don't think this is the crypto structure you want here. The issue is that you have a non-trivial chance of nonce-reuse at 2^96 because of birthday, which is catastrophic with GCM. You should either use an integer nonce (that won't repeat within any reasonable bound) or instead generate a longer nonce (256 bits is good) and use it to derive both the per-message key and per-message nonce